### ISSUE TITLE
Introduce new package model/type in the lockfile to identify custom packages

### ISSUE THREAD

**Issue by billlevine** [flox/flox#2492]
Currently there is a single type to identify a package in the lockfile that came from the catalog `LockedPackageCatalog`.  Now that we are introducing custom packages, we need to be able to identify them.  For now we should infer the difference at the API layer, checking the already provided `catalog` field, and create a corresponding lockfile type.  We will need to also update the logic in `realise_lockfile` to remove the heuristic and add handling for the new type.

It is understood this will not be backwards compatible once you lock with the new code _AND_ have custom packages.  It should _NOT_ break compatibility with existing lockfiles that have no custom packages.

**Comment by limeytexan** [comment]
By "custom packages" do you just mean "flox builds"? I.e. the NEF and manifest builds that require flox to perform the build if it needs to be built at resolution time?

Can we just recap the types of packages?

* flake references
    * of which base catalog entries are essentially a subset because we fall back to building by flakeref if not available from the binary cache
* storepaths
* flox builds; i.e. packages for which you must clone the locked repo and invoke "flox build" to create the package

**Comment by billlevine** [comment]
> By "custom packages" do you just mean "flox builds"? I.e. the NEF and manifest builds that require flox to perform the build if it needs to be built at resolution time?
> 
> Can we just recap the types of packages?
> 
> * flake references
>   
>   * of which base catalog entries are essentially a subset because we fall back to building by flakeref if not available from the binary cache
> * storepaths
> * flox builds; i.e. packages for which you must clone the locked repo and invoke "flox build" to (re)create the package

Correct.  Packages not from our nixpkgs catalog, but from a `flox build`.  I think currently flake references and storepath entries in the lockfile are structurally different to facilitate de-serialization into different data types in Rust.  That is not true for packages from `nixpkgs` versus a user created catalog from a build.  This could be a larger discussion around unifying those models across the platform (API, Db, lockfile, etc) and using content markers to know how to handle them.  When and if that's needed can probably be held off some.
